Added IL2 PARAFAC Analysis Example Script

borcuttjahns opened the pull request:
Needs review from @aarmey and TensorLy maintainers.

aarmey commented:
This looks good to me. We'll need to wait for some input about how the documentation building works.

JeanKossaifi commented on tensorly/datasets/nn_parafac_IL2_example.py:
```diff
@@ -0,0 +1,89 @@
+"""
+Non-negative PARAFAC Decomposition of IL-2 Response Data
+==============================
```
the `=` should go all the way to the end of the actual text.

JeanKossaifi commented on tensorly/datasets/nn_parafac_IL2_example.py:
```diff
@@ -0,0 +1,89 @@
+
+#%%
+# Here we will load a tensor of experimentally measured cellular responses to 
+# IL-2 stimulation. IL-2 is a naturally occurring immune signaling molecule 
```
Might be useful for non-experts to have a slightly longer intro, e.g. what kind of signalling does it do, etc?
This example is super interesting for non experts too!

JeanKossaifi commented on tensorly/datasets/nn_parafac_IL2_example.py:
```diff
@@ -0,0 +1,89 @@
+
+
+tensorFact = non_negative_parafac(np.nan_to_num(responseData.tensor), init='random', rank=3, mask=np.isfinite(responseData.tensor), n_iter_max=5000, tol=1e-9, random_state=1)
+tensorFact.normalize()
```
Perhaps we should call explicitly the normalize function from the module on the factorization rather than use its method?

JeanKossaifi commented on tensorly/datasets/nn_parafac_IL2_example.py:
```diff
@@ -0,0 +1,89 @@
+# future comparisons of correlations across components.
+
+
+tensorFact = non_negative_parafac(np.nan_to_num(responseData.tensor), init='random', rank=3, mask=np.isfinite(responseData.tensor), n_iter_max=5000, tol=1e-9, random_state=1)
```
for instances, i'd use a `tensor_fact`, or better, a more explicit naming.

JeanKossaifi reviewed (commented):
Thanks @borcuttjahns and @aarmey, this will be great to have, and super interesting example!

The examples use the sphinx gallery : the file should be in the /examples folder in the repository, and the name start with `plot_` and it should be automatically executed at each build of the doc.

aarmey commented:
Thanks for the input, @JeanKossaifi.

JeanKossaifi commented:
There's a conflict with main as we changed a little the structure of datasets.

borcuttjahns commented:
Alright - I think I addressed the comments both on this PR as well as those mentioned in the meeting regarding dataset structure. While I still am a bit unsure of how to preview Sphinx Gallery builds, I believe this PR might be ready for another check?

aarmey reviewed (approved):
This looks right to me.

JeanKossaifi commented:
@borcuttjahns to preview the example, you can build the doc
```bash
cd doc
pip install -r requirements_doc.txt
make html
```
you can then open _build/html/index.html to view the local doc

JeanKossaifi commented on doc/Makefile:
```diff
@@ -5,7 +5,7 @@
 SPHINXOPTS    =
 SPHINXBUILD   = sphinx-build
 PAPER         =
-BUILDDIR      = _build
```
Why did you change this? I will break the Github scripts to automatically build the doc.

JeanKossaifi commented:
Looks good to me - it's still marked as draft, do you have more changes to make?

borcuttjahns commented:
> Looks good to me - it's still marked as draft, do you have more changes to make?

Should be good to go! Sorry I missed notifying you after the last update - but it looks to be building correctly with Sphinx.

JeanKossaifi commented:
Sorry for the delay - it looks good to me! Can we just move it to the applications folder in examples, along with the Covid one?

borcuttjahns commented:
Sounds good - just moved it! Just let me know if there are any other changes you'd like me to make before this gets merged up.

JeanKossaifi reviewed (commented):
Looks great, sorry for the delay in reviewing!
Happy to merge, there's just a change in the sphinx doc Makefile that I think would break the CI workflow.

aarmey commented:
@JeanKossaifi I reverted that change. It looks ready to merge to me.

JeanKossaifi commented:
Perfect, thanks! Merging.

JeanKossaifi commented:
Perfect, thanks! Merging.

JeanKossaifi commented:
Ok, I need to add a dummy documentation build to the PR tests.  Seems like we broke the documentation, here are the errors:
```
generating gallery for auto_examples/applications... [ 33%] plot_image_compression.py
generating gallery for auto_examples/applications... [ 66%] plot_IL2.py
WARNING: /home/runner/work/tensorly/tensorly/examples/applications/plot_IL2.py failed to execute correctly: Traceback (most recent call last):
  File "/home/runner/work/tensorly/tensorly/examples/applications/plot_IL2.py", line 14, in <module>
    import pandas as pd
ModuleNotFoundError: No module named 'pandas'

generating gallery for auto_examples/applications... [100%] plot_covid.py
WARNING: /home/runner/work/tensorly/tensorly/examples/applications/plot_covid.py failed to execute correctly: Traceback (most recent call last):
  File "/home/runner/work/tensorly/tensorly/examples/applications/plot_covid.py", line 36, in <module>
    from tensorly.datasets.imports import load_covid19_serology
ModuleNotFoundError: No module named 'tensorly.datasets.imports'
```

JeanKossaifi commented:
Can we do without pandas or do we need to make it a dependency, at least for the doc? The other issue is a quick fix.

JeanKossaifi commented:
@borcuttjahns @aarmey can we easily run the example without pandas or do we need to add it to (at least the doc's) dependencies?

JeanKossaifi commented:
Apologies, I had missed it, thanks, I've merged it! :)
